File: intersection_two_linkedlists.py
# Time Complexity : O(M * N) where M and N is number of nodes in the two linked lists respectively.
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        curr = headA
        i = 1
        while curr.next is not None:
            curr = curr.next
            i += 1
            
        curr = headB
        j = 1
        while curr.next is not None:
            curr = curr.next
            j += 1
                        
        if i > j:
            diff = i - j
            currA = headA
            currB = headB
            while diff > 0:
                diff -= 1
                currA = currA.next
            
            while currB != currA and currA is not None and currB is not None:
                currA = currA.next
                currB = currB.next
                
            return currA
        else:
            diff = j - i
            currA = headA
            currB = headB
            while diff > 0:
                diff -= 1
                currB = currB.next
            
            while currB != currA and currA is not None and currB is not None:
                currA = currA.next
                currB = currB.next
                
            return currA

File: IntersectionTwoLinkedLists.java
// Time Complexity : O(m+n)
// Spcae Complexity : O(1)
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class IntersectionTwoLinkedLists {
    public static void main(String[] args) {
        ListNode headA = new ListNode(4);
        headA.next = new ListNode(1);
        headA.next.next = new ListNode(8);
        headA.next.next.next.next = new ListNode(4);
        headA.next.next.next.next.next = new ListNode(5);
        ListNode headB = new ListNode(5);
        headB.next = new ListNode(6);
        headB.next.next = new ListNode(1);
        headB.next.next.next = new ListNode(8);
        headB.next.next.next.next = new ListNode(4);
        headB.next.next.next.next.next = new ListNode(5);
        ListNode intersectionPoint = getIntersectionNode(headA, headB);
        System.out.println(intersectionPoint);
    }

    private static ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null)
            return null;
        ListNode currA = headA;
        int lenA = 0;
        while (currA != null) {
            lenA++;
            currA = currA.next;
        }
        ListNode currB = headB;
        int lenB = 0;
        while (currB != null) {
            lenB++;
            currB = currB.next;
        }
        while (lenA > lenB) {
            headA = headA.next;
            lenA--;
        }
        while (lenB > lenA) {
            headB = headB.next;
            lenB--;
        }
        while (headA != headB) {
            headA = headA.next;
            headB = headB.next;
        }
        return headA;
    }
}